Language/Classes/Instance_Methods/Method_noSuchMethod/forwarder_A01_t05.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Let C be a concrete class, let L be the library that contains the
+/// declaration of C, and let m be a name. Then m is noSuchMethod forwarded in C
+/// iff one of the following is true:
+///
+/// • Requested in program: C has a non-trivial noSuchMethod, the interface of C
+/// contains a member signature S named m, and C has no concrete member named m
+/// and accessible to L that correctly overrides S (that is, no member named m
+/// is declared or inherited by C, or one is inherited, but it does not have the
+/// required signature). In this case we also say that S is noSuchMethod
+/// forwarded.
+///
+/// • Forced by privacy: There exists a direct or indirect superinterface D of C
+/// which is declared in a library L2 different from L, the interface of D
+/// contains a member signature S named m, m is a private name, and no
+/// superclass of C has a concrete member named m accessible to L2 that
+/// correctly overrides S. In this case we also say that S is noSuchMethod
+/// forwarded.
+///
+/// For a concrete class C, a noSuchMethod forwarder is implicitly induced for
+/// each member signature which is noSuchMethod forwarded.
+///
+/// @description Checks that for a concrete class, a `noSuchMethod` forwarder is
+/// implicitly induced for each member signature which is `noSuchMethod`
+/// forwarded. Test the case when forwarding is requested in program and `C` has
+/// no concrete member named `m`
+/// @author sgrekhov22@gmail.com
+
+import "no_such_method_lib.dart";
+import "../../../../Utils/expect.dart";
+
+class C1 extends D1 {
+  String _m();
+}
+
+class C2 extends D2 {
+  String _m();
+  dynamic noSuchMethod(Invocation i) => "C2";
+}
+
+mixin M1 on D1 {
+  String _m();
+}
+
+mixin M2 on D2 {
+  String _m();
+  dynamic noSuchMethod(Invocation i) => "M2";
+}
+
+class MA1 = D1 with M1;
+class MA2 = D2 with M2;
+
+enum E1 with D1 {
+  e1, e2;
+  String _m();
+}
+
+enum E2 with D2 {
+  e1, e2;
+  String _m();
+  dynamic noSuchMethod(Invocation i) => "E2";
+}
+
+main() {
+  Expect.equals("D1", C1()._m());
```

Comment:
This kind of test would actually succeed even if an implementation of Dart would use the old approach (it would then detect that there is no `_m` member of the given object `C1()`, and it would invoke `noSuchMethod` using the actual arguments).

So we need to do something which is a bit more circumspect in order to ensure that there _is_ a noSuchMethod forwarder.

One thing we can do is to tear it off: If we had that old semantics and invoke `C1()._m` then it would invoke `noSuchMethod` and we would receive `"C1"`, which would then probably give rise to a dynamic type error because `C1()._m` should yield a function object.

So we'd tear off the method, then test that it is a function object with the expected type (`is String Function()`), and then we'd know that there _is_ a real member named `_m` in the object, and that must be the noSuchMethod forwarder. We could then call the function object and see that it returns `"C1"`, such that we can see that it works as expected.

Reply:
With the old semantics it was necessary to write code in `noSuchMethod` to handle the case where the given `Invocation` is a getter invocation (that's a tear-off operation, so we should return a function object) and when it's a method invocation (that's the case where we'd use the behavior that we have today), so it _was_ possible to get the same behavior.

However, in this case we know what's in the body of the relevant `noSuchMethod`, so if the tear-off works then we also know that there is a noSuchMethod forwarder.